picard/util/pipe.py

```diff
@@ -0,0 +1,241 @@
+#
+# Copyright (C) 2022 skelly37
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+
+import concurrent.futures
+import os
+from typing import Optional
+
+from picard.const.sys import (
+    IS_MACOS,
+    IS_WIN,
+)
+
+
+if IS_WIN:
+    import win32pipe  # type: ignore
+    import win32file  # type: ignore
+    from pywintypes import error as WinApiError  # type: ignore
+
+
+class Pipe:
+    NO_RESPONSE_MESSAGE: str = "No response from FIFO"
+    NOT_FOUND_MESSAGE: str = "FIFO doesn't exist"
+    MESSAGE_TO_IGNORE: str = "Ignore this message, just testing the pipe"
+    TIMEOUT_SECS: float = 1.5
+
+    def __init__(self, app_name: str, app_version: str, args=None):
+        if args is None:
+            args = []
+
+        self.__app_name: str = app_name
+        self.__app_version: str = app_version
+        self.__is_mac: bool = IS_MACOS
+        self.__is_win: bool = IS_WIN
+
+        # named pipe values needed by windows API
+        if self.__is_win:
+            self.__app_version = self.__app_version.replace(".", "-")
+
+            # win32pipe.CreateNamedPipe
+            # more about the arguments: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createnamedpipea
+            self.__MAX_INSTANCES: int = 1
+            self.__BUFFER_SIZE: int = 65536
+            # timeout doesn't really matter, concurrent.futures ensures that connections are closed in declared time
+            # the value is in milliseconds
+            self.__DEFAULT_TIMEOUT: int = 300
+
+            # win32file.CreateFile
+            # more about the arguments: http://timgolden.me.uk/pywin32-docs/win32file__CreateFile_meth.html
+            self.__SHARE_MODE: int = 0
+            self.__FLAGS_AND_ATTRIBUTES: int = 0
+
+            # pywintypes.error error codes
+            # more about the error codes: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/18d8fbe8-a967-4f1c-ae50-99ca8e491d2d
+            self.__FILE_NOT_FOUND_ERROR_CODE: int = 2
+            self.__BROKEN_PIPE_ERROR_CODE: int = 109
+
+        self.path: str = self.__generate_filename()
+
+        self.is_pipe_owner: bool = False
+        self.permission_error_happened: bool = False
+
+        # test if pipe is listened to even if no args provided
+        if isinstance(args, list):
```

Comment:
Instead we could try a conversion to a list, so any  iterable can be passed as `args`.

Also, it seems this list is only modified here, so it could be made a tuple instead.

Something like:

```python
try:
    args = tuple(args)
    if not args:
        args = (self.MESSAGE_TO_IGNORE, )
except TypeError as exc:
    raise PipeErrorBadArgs(exc)
```

Reply:
Such conversion might raise misbehavior, like:

```python
>>> tuple("hello")
('h', 'e', 'l', 'l', 'o')
```

I'd rather explicitly declare allowed types, like:

```python
if type(args) in [list, set]:
    args = tuple(args)

if isinstance(args, tuple):
    if not args:
        args = (self.MESSAGE_TO_IGNORE, )
else:
    raise PipeErrorBadArgs()
```

Reply:
Okay, check out the current commit how I've solved it.